doc/dev/kclient.rst

```diff
@@ -0,0 +1,345 @@
+Testing changes to the Linux Kernel CephFS driver
+=================================================
+
+This walkthrough will explain one (opinionated) way to do testing of the Linux
+kernel client against a development cluster. We will try to mimimize any
+assumptions about pre-existing knowledge of how to do kernel builds or any
+related best-practices.
+
+Step One: build the kernel
+==========================
+
+Configure the kernel:
+
+.. bash::
+
+    make defconfig
+
+
+We now have a kernel config with reasonable defaults for the architecture you're
+building on. The next thing to do is to enable configs which will build Ceph and/or
+provide functionality we need to do testing.
+
+.. bash::
+
+    cat > ~/.ceph.config <<EOF
+    CONFIG_XFS_FS=y
+    CONFIG_CEPH_FS=y
+    CONFIG_CEPH_FSCACHE=y
+    CONFIG_CEPH_FS_POSIX_ACL=y
+    CONFIG_CEPH_FS_SECURITY_LABEL=y
+    CONFIG_CEPH_LIB_PRETTYDEBUG=y
+    CONFIG_DYNAMIC_DEBUG=y
+    CONFIG_DYNAMIC_DEBUG_CORE=y
+    CONFIG_KGDB=y
+    CONFIG_FRAME_POINTER=y
+    CONFIG_KGDB_SERIAL_CONSOLE=y
+    EOF
+
+Note, beyond enabling Ceph-related configs, we are also enabling some useful
+debug configs and XFS (as an alternative to ext4 if needed for our root file
+system).
+
+Now, merge the configs.
```

Comment:
Why not just generated the `.config` from https://github.com/lxbsz/ceph-build ?

For example, on RHEL/Fedora/Centos-stream we can generate the config file from https://github.com/lxbsz/ceph-build/blob/18e084ed913a9ebd13c37bfde24b170fa65afdd5/kernel/build/kernel-config-rpm.sh.

This will include all the configs needed for cephfs kclient, such as the ENCRYPT, FSCACHE, etc.


For the fscrypt feature testing you can refer to Jeff's doc https://jtlayton.wordpress.com/2021/11/29/testing-the-linux-kernel-cephfs-client-with-xfstests/.

Reply:
> Why not just generated the `.config` from https://github.com/lxbsz/ceph-build ?

It's a good question. I think for general local development many of those configs may not be necessary and slow down builds but I haven't tried.

> For example, on RHEL/Fedora/Centos-stream we can generate the config file from https://github.com/lxbsz/ceph-build/blob/18e084ed913a9ebd13c37bfde24b170fa65afdd5/kernel/build/kernel-config-rpm.sh.
> 
> This will include all the configs needed for cephfs kclient, such as the ENCRYPT, FSCACHE, etc.

Sure, that those in particular should be added to this doc (but they're not strictly necessary).

> For the fscrypt feature testing you can refer to Jeff's doc https://jtlayton.wordpress.com/2021/11/29/testing-the-linux-kernel-cephfs-client-with-xfstests/.

I'll try to incorporate parts of that. Thanks for the pointer.

Reply:
Yeah, certainly you can do it manually, which is what I did, too if you know which configs needed to enable. But for a newbee this script is a good start.